src/write/line.rs

```diff
@@ -1003,177 +1018,378 @@ mod convert {
-            line_strings: &mut write::LineStringTable,
-            strings: &mut write::StringTable,
-        ) -> ConvertResult<LineString> {
-            Ok(match from_attr {
-                read::AttributeValue::String(r) => LineString::String(r.to_slice()?.to_vec()),
-                read::AttributeValue::DebugStrRef(offset) => {
-                    let r = dwarf.debug_str.get_str(offset)?;
-                    let id = strings.add(r.to_slice()?);
-                    LineString::StringRef(id)
-                }
-                read::AttributeValue::DebugLineStrRef(offset) => {
-                    let r = dwarf.debug_line_str.get_str(offset)?;
-                    let id = line_strings.add(r.to_slice()?);
-                    LineString::LineStringRef(id)
+        /// Read the next sequence from the source program.
+        pub fn read_sequence(&mut self) -> ConvertResult<Option<LineConvertSequence>> {
+            let mut start = None;
+            let mut rows = Vec::new();
+            while let Some(row) = self.read_row()? {
+                match row {
+                    LineConvertRow::SetAddress(address, row) => {
+                        // We only support the setting the address for the first row in a sequence.
+                        if !rows.is_empty() {
+                            return Err(ConvertError::InvalidAddress);
+                        }
+                        start = Some(address);
+                        rows.push(row);
+                    }
+                    LineConvertRow::Row(row) => {
+                        rows.push(row);
+                    }
+                    LineConvertRow::EndSequence(length) => {
+                        return Ok(Some(LineConvertSequence {
+                            start,
+                            length,
+                            rows,
+                        }));
+                    }
                 }
-                _ => return Err(ConvertError::UnsupportedLineStringForm),
-            })
+            }
+            Ok(None)
+        }
+
+        /// Call [`LineProgram::begin_sequence`] for the converted program.
+        pub fn begin_sequence(&mut self, address: Option<Address>) {
+            self.program.begin_sequence(address);
+        }
+
+        /// Call [`LineProgram::end_sequence`] for the converted program.
+        pub fn end_sequence(&mut self, address_offset: u64) {
+            self.program.end_sequence(address_offset);
+        }
+
+        /// Return [`LineProgram::in_sequence`] for the converted program.
+        pub fn in_sequence(&self) -> bool {
+            self.program.in_sequence()
+        }
+
+        /// Set the next row and call [`LineProgram::generate_row`] for the converted program.
+        pub fn generate_row(&mut self, row: LineRow) {
+            *self.program.row() = row;
+            self.program.generate_row();
+        }
+
+        /// Return the program and a mapping from source file index to `FileId`.
```

Comment:
Maybe mention the subtelties with `files[0]` for dwarf <=4 here, and how a default entry is added.